### ISSUE TITLE
Sampling from flat hyperrectangle fails

### ISSUE THREAD

**Issue by mforets** [JuliaReach/LazySets.jl#2058]
```julia
julia> using LazySets, Distributions
[ Info: Precompiling Distributions [31c24e10-a181-5473-b8eb-7969acd0382f]

julia> H = Hyperrectangle(low=[0.0, 0.0], high=[0.0, 1.0])
Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}([0.0, 0.5], [0.0, 0.5])

julia> isflat(H)
true

julia> sample(H, 10)
WARNING: both Distributions and LazySets export "sample"; uses of it in module Main must be qualified
ERROR: UndefVarError: sample not defined
Stacktrace:

 [1] top-level scope at REPL[4]:1

julia> LazySets.sample(H, 10)
ERROR: ArgumentError: Uniform: the condition a < b is not satisfied.
Stacktrace:
 [1] macro expansion at /home/mforets/.julia/packages/Distributions/2i7UF/src/utils.jl:6 [inlined]
 [2] #Uniform#104 at /home/mforets/.julia/packages/Distributions/2i7UF/src/univariate/continuous/uniform.jl:33 [inlined]
 [3] Uniform(::Float64, ::Float64) at /home/mforets/.julia/packages/Distributions/2i7UF/src/univariate/continuous/uniform.jl:33
 [4] collect(::Base.Generator{UnitRange{Int64},LazySets.var"#308#312"{UnionAll,Array{Float64,2}}}) at ./none:0
 [5] LazySets.RejectionSampler(::Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}, ::Type) at /home/mforets/.julia/dev/LazySets/src/Utils/samples.jl:231
 [6] RejectionSampler at /home/mforets/.julia/dev/LazySets/src/Utils/samples.jl:228 [inlined]
 [7] #sample#262(::Nothing, ::Random._GLOBAL_RNG, ::Nothing, ::typeof(LazySets.sample), ::Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}, ::Int64) at /home/mforets/.julia/d
ev/LazySets/src/Utils/samples.jl:55
 [8] sample(::Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}, ::Int64) at /home/mforets/.julia/dev/LazySets/src/Utils/samples.jl:47
 [9] top-level scope at REPL[5]:1

julia> H = Hyperrectangle(low=[0.0, 0.0], high=[0.01, 1.0])
Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}([0.005, 0.5], [0.005, 0.5])

julia> LazySets.sample(H, 10)
10-element Array{Array{Float64,1},1}:
 [0.004809780755984916, 0.8224446314030849]
 [0.004506972704302776, 0.8169792624939511]
 [0.002170279929780925, 0.5435194729471067]
 [0.009953623375562394, 0.5327266054879436]
 [0.0009795361769845078, 0.016023126258301046]
 [0.006065078107734037, 0.8561376509082252]
 [0.005979859938608933, 0.9697289514576151]
 [0.007033106439952292, 0.2191717860753264]
 [0.002565670539134497, 0.4886645392072382]
 [0.0002322563631416297, 0.7572548970270925]
```

**Comment by schillic** [comment]
The problem is that a distribution requires a dense set to sample from.

https://github.com/JuliaReach/LazySets.jl/blob/894f77bd2309a80e02d5be50acfb6dbc7c617947/src/Utils/samples.jl#L231

We use
https://github.com/JuliaReach/LazySets.jl/blob/894f77bd2309a80e02d5be50acfb6dbc7c617947/src/Utils/samples.jl#L260
to sample from this distribution later.

Not sure how to solve this. A hacky workaround is to remove flat dimensions, remember their index, sample a point from the reduced state space, and finally construct the result vector by inserting the constant values for the flat dimensions.

**Comment by schillic** [comment]
An unsound but practical workaround is to sample from a bloated set instead. Unfortunately, `sample(Bloating(X, 1e-10))` does not work because we do not offer a membership test.

**Comment by mforets** [comment]
> remove flat dimensions, remember their index, sample a point from the reduced state space, and finally construct the result vector by inserting the constant values for the flat dimensions.

i think that's a reasonable thing to do. here is an implementation:

```julia
function sample_from_flat(H::AbstractHyperrectangle, num_samples=1)
    @assert isflat(H)
    r = radius_hyperrectangle(H)
    c = center(H)
    n = length(r)

    zind = iszero.(r)
    nzind = @. !zind
    rnz = r[nzind]
    cnz = c[nzind]
    Hnz = Hyperrectangle(cnz, rnz)
    samples = sample(Hnz, num_samples)

    out = [copy(c) for _ in 1:num_samples]
    for (s, o) in zip(samples, out) 
        view(o, nzind) .= s
    end
    return out
end

**Comment by mforets** [comment]
this works in current master, so should we close it?

```julia
julia> using LazySets
[ Info: Precompiling LazySets [b4f0291d-fe17-52bc-9479-3d1a343d9043]

julia> H = Hyperrectangle(low=[0.0, 0.0], high=[0.0, 1.0])
Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}([0.0, 0.5], [0.0, 0.5])

julia> isflat(H)
true

julia> sample(H, 10)
10-element Array{Array{Float64,1},1}:
 [0.0, 0.4128312789713371]
 [0.0, 0.004127705473208287]
 [0.0, 0.3930197267829645]
 [0.0, 0.6606811567219335]
 [0.0, 0.13854748042062615]
 [0.0, 0.6011601227406187]
 [0.0, 0.7134655881282674]
 [0.0, 0.47706237284784114]
 [0.0, 0.07518756137640481]
 [0.0, 0.8975219101722374]
```

**Comment by schillic** [comment]
This was fixed in #2480.
